Add option to reinitialise cache when cleaning or flushing

arollason opened the pull request:
Magerun pull-request check-list:

- [x] Pull request against develop branch (if not, just close and create a new one against it)
- [x] README.md reflects changes (if any)

Subject: Add option to reinitialise cache when cleaning or flushing

Previously the config cache would always be reinitialised when cleaning
or flushing. This was done to resolve a corner case issue with the
changing of database credentials. As the majority of the time no
reinitialisation is required, we move it into a --reinit option, making
cache cleaning much faster, and less prone to error.
    
It should be noted that the method used for reinitialisation of the
config cache will reliably corrupt the config cache when using redis.
There is a "reinit" method in the Mage_Core_Model_Config class which
will reinitialise the config cache and not cause any corruption.However,
instead of changing the method used to rebuild the config cache, this
patch simply puts the automatic rebuild behind an optional flag, as the
config cache reinitialisation is not required when clearing other caches.

tkn98 commented:
Dear @arollason, thanks a lot for your pull request.

Genereally about the feature you offer here: I can understand you're not confident with the defaults. However we can not change the previous behavior to be backwards compatible.

What I can offer is that we have two flags `--reinit` and `--no-reinit`. The first switch is implicit and technically not needed however allows to override the second one if given so that we can think about a future (if we don't have it yet) local config override for magerun command switch which will give you the outcome you're aiming for while we're maintaining backwards compatibility.

So please preserve the current default behavior and add another switch so that both ways are possible.

arollason commented:
Hey @tkn98, Thanks for the quick response. It all sounds reasonable. I'll amend the PR on Monday with the suggested changes. Cheers.

tkn98 commented:
@arollason: Thanks a lot, just take your time. If you need any feedback or assistance just comment here and we can see what we can do.

arollason commented:
I've updated the pull request to include the no-reinit option. I removed the calls to "hasOption" because they always returned true. getOption returns the expected value of whether the flag is present in the command or not. If --reinit and --no-reinit are both used in the same command, a reinitialisation of the config cache will be run. If no flags are provided a reinitialisation will be run. If only --no-reinit is provided, no reinitialisation will be run.
Let me know if any other changes are required.

tkn98 commented:
I'd say as reinitialisation is the default we currently have, if you specify both switches (and as we can't determine which comes last IIRC from Symfony console) --no-reinit should win if both switches are used.

Also please squash your commits into one for this feature so I can take a better view on the code at once. You can force-push the feature branch and this pull-request will automatically be updated.

arollason commented:
If that's the desired functionality then it probably makes sense to just remove the --reinit flag entirely, as it won't be doing anything or need to be checked in the code:

cache:flush -> Flush with a config reinit
cache:flush --reinit -> Flush with a config reinit
cache:flush --no-reinit -> Flush without config reinit
cache:flush --reinit --no-reinit -> Flush without config reinit

This way the check can simply be:

if (!$input->getOption('no-reinit')) {
  // do config cache reinit
}

And there's no reason to even include --reinit in the documentation or code, other than documenting that a config reinitialisation will occur by default when running the cache:flush and cache:clean commands. If that sounds reasonable I can update and squash the PR with the change.

tkn98 commented:
The `--reinit` would allow to create portable scripts in the future when we change the default behavior.

arollason commented:
Ok no worries, I've pushed up the change that leaves the flag --reinit as an available option, and the config cache reinitialisation will occur by default. If --no-reinit is passed no reinitialisation will occur.

ktomk commented on readme.rst:
```diff
@@ -685,13 +685,17 @@ If you would like to clean only one cache type:
 .. code-block:: sh
 
-   $ n98-magerun.phar cache:clean [code]
+   $ n98-magerun.phar cache:clean [code] [--reinit --no-reinit]
```
I think this should be:

     $ n98-magerun.phar cache:clean [--reinit][--no-reinit][<code>]

This takes care about the two new options and fixes the misnomer on the operand. I'm aware this is not yet consistent style, so perhaps just for the two new options and you might want to separate by space. Please decide on your own.

ktomk commented on readme.rst:
```diff
@@ -685,13 +685,17 @@ If you would like to clean only one cache type:
 .. code-block:: sh
 
-   $ n98-magerun.phar cache:clean [code] [code] ...
+   $ n98-magerun.phar cache:clean [code] [code] ... [--reinit --no-reinit]
```
Similar here, please place the options before the operands.

ktomk commented on readme.rst:
```diff
@@ -685,13 +685,17 @@ If you would like to clean only one cache type:
+
+Options:
+    --reinit Reinitialise the config cache after cleaning (Default)
+    --no-reinit Don't reinitialise the config cache after cleaning
```
A short notice about that --reinit overrides --no-reinit would be fine.

ktomk commented on readme.rst:
```diff
@@ -700,9 +704,15 @@ Run `cache:list` command to see all codes.
 .. code-block:: sh
 
-   $ n98-magerun.phar cache:flush
+   $ n98-magerun.phar cache:flush [--reinit --no-reinit]
```
ibid.

ktomk commented on src/N98/Magento/Command/Cache/CleanCommand.php:
```diff
@@ -29,6 +42,9 @@ protected function configure()
 
+Options:
+    --reinit Reinitialise the config cache after cleaning (Default)
+    --no-reinit Don't reinitialise the config cache after cleaning
```
ibid.

tkn98 commented:
Review usage manual interaction failure. The Github U/I is to be learned first. @arollason Please see the comments in the reivew, looks like I clicked once too often last night.

arollason commented:
Hi ktomk, thanks for taking the time to review. I've updated the documentation to reflect that --no-reinit will override --reinit. I used spaced to separate the options to be consistent with the other commands with options in the documentation already.

tkn98 commented:
Thanks! This is now part of the development version will go into next stable.
